ql/types.hpp

```diff
@@ -81,6 +81,12 @@ namespace QuantLib {
     /*! \ingroup types */
     typedef Real Probability;
 
+    //! type conversions wrapper Real -> T, to allow specialisations for custom Real types
+    /*! \ingroup types */
+    template <class T, class F>
```

Comment:
According to your comment, the intended use case for this function is for converting `Real` to another type `T`. If that is so then I recommend enforcing this usage like so:

`template <class T, class F, class = typename std::enable_if<std::is_same<F, Real>::value>::type>`

Also in this case since `Real` should be cheap to copy I recommend passing it by value instead of by reference.

Reply:
We considered this, and actually had it without the second template parameter in the first place (just taking a `Real` argument directly). However it creates an issue in a template context, where the type it works with may or may not be `Real` and the `ql_cast` call must work for any type.

 An example is here https://github.com/xcelerit/QuantLib/blob/feature/ql_cast/ql/time/period.hpp#L148 .

Reply:
In that case, perhaps a reasonable compromise would be to require any numerical type instead of just `Real`.

`template <class T, class F, class = typename std::enable_if<std::is_arithmetic<F>::value>::type>`

Also, just curious why the `Real` type you're planning to use can't define its own conversions to integral types? Is it a language problem or a library problem or something else?

Reply:
Initially, we thought about this as a safety feature, as for some `Real` types converting to integer by accident and then back to `Real` would loose some of the features of the custom `Real` type. However, we have confirmed that the current use of these Real->int casts in QuantLib are safe and an automatic conversion operator would be fine. 

Pending your and @lballabio confirmation, we can go ahead and cancel this PR.

Reply:
Fine with me to close the PR if no longer required.